src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,537 @@
+#include <stdint.h>
+
+static const struct comp_driver comp_igo_nr;
+
+/* 696ae2bc-2877-11eb-adc1-0242ac120002 */
+DECLARE_SOF_RT_UUID("igo-nr", igo_nr_uuid,  0x696ae2bc, 0x2877, 0x11eb, 0xad, 0xc1,
+		    0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(igo_nr_tr, SOF_UUID(igo_nr_uuid), LOG_LEVEL_INFO);
+
+static void igo_nr_capture_s16(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int16_t *x;
+	int16_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s16(source, idx_in);
+		cd->in[cd->in_wpt] = (*x);
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]);
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]);
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static void igo_nr_capture_s32(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int32_t *x;
+	int32_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = (*x) >> 16;
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]) << 16;
+
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]) << 16;
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+
+static inline int32_t set_capture_func(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	switch (sourceb->stream.rate) {
+	case 16000:
+		comp_info(dev, "set_capture_func(), sample rate = 16000");
+		break;
+	default:
+		comp_err(dev, "set_capture_func(), invalid sample rate");
+		return -EINVAL;
+	}
+
+	switch (sourceb->stream.frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S16_LE");
+		cd->igo_nr_func = igo_nr_capture_s16;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S32_LE");
+		cd->igo_nr_func = igo_nr_capture_s32;
+		break;
+	default:
```

Comment:
Do you have plan to support S24_4LE? It's quite straightforward with shifts like done with S32_LE.

Reply:
sure we can. will update then.